File: sort/InsertionSort.php
<?php
/**
 * Created by PhpStorm.
 * User: alexander
 * Date: 08.03.18
 * Time: 07:58
 */

namespace Alex\Sort;


class InsertionSort implements Sorter
{
    public function __invoke(array $sortables): array
    {
        $output = array();
        $lengthOfOutput = 0;
        foreach ($sortables as $element) {
            $max = $lengthOfOutput;
            while (0 < $max && $output[$max - 1] > $element) {
                $output[$max] = $output[$max - 1];
                $max--;
            }

            $output[$max] = $element;
            $lengthOfOutput++;
        }

        return $output;
    }
}

File: sort/go/insertion_sort.go
package sort

func Insertion_sort(input *[]int64) {
	sortables := *input
	*input = sort_insertion(sortables)
}

func sort_insertion(input sortables) sortables {
	length := len(input)
	for i := 1; i < length; i++ {
		for j := i; j > 0; j-- {
			if input[j] < input[j-1] {
				tmp := input[j]
				input[j] = input[j-1]
				input[j-1] = tmp
			}
		}
	}
	
	return input
}